### PR TITLE
Refactor reconciler logic

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
## Description

I had refactor the all reconciliation loops of **odh-model-controller** by following below guideline:
https://docs.google.com/document/d/1Q7ziQa_DtqtHnP9oWKNVgWNboYL0LhkdsHXl43QtXqI/edit?usp=sharing

Refactoring would fix the https://github.com/opendatahub-io/odh-model-controller/issues/86 issue.

## How Has This Been Tested?
- Create a new cluster
- Setup the cluster with below Installation step
https://github.com/opendatahub-io/caikit-tgis-serving/blob/main/demo/kserve/install-manual.md
- Before executing Step 11, specify custom manifests
```
apiVersion: dscinitialization.opendatahub.io/v1
kind: DSCInitialization
metadata:
  name: example
spec:
  devFlags:
    manifestsUri: https://github.com/vaibhavjainwiz/odh-manifests/tarball/refactor
```
- Apply Step 11.  
- Deploy/validate a Kserve model with below Step.
https://github.com/opendatahub-io/caikit-tgis-serving/blob/main/demo/kserve/deploy-remove.md

If everything work as mentioned in above step then we are good to go :)

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/comparators/peerAuthentication_comparator.go
Comment (spolti) [inline controllers/comparators/peerAuthentication_comparator.go:15]: Are these the only fields needed to compare?

File: controllers/comparators/route_comparator.go
Comment (spolti) [inline controllers/comparators/route_comparator.go:13]: isn't better to compare the Route itself than its fields?

File: controllers/comparators/service_comparator.go
Comment (spolti) [inline controllers/comparators/service_comparator.go:13]: same here

**Review by spolti — COMMENTED** [review]
There are some files missing the license headers while others have.
Do we want to keep it aligned?

File: controllers/comparators/serviceaccount_comparator.go
Comment (spolti) [inline controllers/comparators/serviceaccount_comparator.go:9]: Are you planning to add some logic here?

File: controllers/reconcilers/mm_serviceaccount_reconciler.go
Comment (spolti) [inline controllers/reconcilers/mm_serviceaccount_reconciler.go:69]: don't we need to to a reconcile.after here?

File: controllers/reconcilers/mm_route_reconciler.go
Comment (spolti) [inline controllers/reconcilers/mm_route_reconciler.go:91]: why not parse to boolean?

File: controllers/reconcilers/kserve_networkpolicy_reconciler.go
Comment (spolti) [inline controllers/reconcilers/kserve_networkpolicy_reconciler.go:64]: is it a good idea to hardcode it?

File: controllers/reconcilers/kserve_inferenceservice_reconciler.go
Comment (spolti) [inline controllers/reconcilers/kserve_inferenceservice_reconciler.go:133]: why can't re rely on the owner metadata information to delete everything?

File: controllers/comparators/peerAuthentication_comparator.go
Comment (vaibhavjainwiz) [reply controllers/comparators/peerAuthentication_comparator.go:15]: My thought process behind field wise comparison is to just compare those fields for which we are responsible to manage. If we compare complete object spec then it imposes unnecessary constraints to the user by not allowing him to make any additional modification to it.
Like in the above spec, Our operator is responsible for defining the `Selector`, `Mtls`, `PortLevelMtls` fields. Apart from these field users could make any modification to `PeerAuthentication` like they could add any labels, annotation etc and we simply don't care.

File: controllers/comparators/serviceaccount_comparator.go
Comment (vaibhavjainwiz) [reply controllers/comparators/serviceaccount_comparator.go:9]: no.. for now we just create a blank ServiceAccount object and didn't set any Spec field. Thats why I directly return true here

File: controllers/reconcilers/kserve_networkpolicy_reconciler.go
Comment (vaibhavjainwiz) [reply controllers/reconcilers/kserve_networkpolicy_reconciler.go:64]: With this PR I just mean to restructure code, even I am not sure why its hard coded here.

File: controllers/reconcilers/kserve_inferenceservice_reconciler.go
Comment (vaibhavjainwiz) [reply controllers/reconcilers/kserve_inferenceservice_reconciler.go:133]: Some of the resources are not directly dependent on instance of InferenceResource.For example IstioSMMR, PrometheusRoleBinding, IstioTelemetry, IstioServiceMonitor, IstioPodMonitor, IstioPeerAuthentication and NetworkPolicy resources are supposed to be created only once for a namespace that why we didn't set the Owner references to them and only delete them when all InferenceResource resources are deleted from that namespace.

File: controllers/comparators/route_comparator.go
Comment (vaibhavjainwiz) [reply controllers/comparators/route_comparator.go:13]: same as above

File: controllers/reconcilers/mm_serviceaccount_reconciler.go
Comment (vaibhavjainwiz) [reply controllers/reconcilers/mm_serviceaccount_reconciler.go:69]: sorry I didn't get it. Could you please explain it more.

Comment (spolti) [reply controllers/reconcilers/mm_serviceaccount_reconciler.go:69]: reconcile.after to set the time until the next reconcile cycle.

File: controllers/reconcilers/kserve_networkpolicy_reconciler.go
Comment (spolti) [reply controllers/reconcilers/kserve_networkpolicy_reconciler.go:64]: okay, maybe we can open a issue to track it later.

File: controllers/reconcilers/kserve_inferenceservice_reconciler.go
Comment (spolti) [reply controllers/reconcilers/kserve_inferenceservice_reconciler.go:133]: I see, thanks for the explanation.

File: controllers/reconcilers/reconciler.go
Comment (davidesalerno) [inline controllers/reconcilers/reconciler.go:1]: What do you think about leaving this interface and change it to be:
`Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error`

File: controllers/reconcilers/mm_serviceaccount_reconciler.go
Comment (vaibhavjainwiz) [reply controllers/reconcilers/mm_serviceaccount_reconciler.go:69]: We have below ticket for same
https://github.com/opendatahub-io/odh-model-controller/issues/89

File: controllers/reconcilers/kserve_networkpolicy_reconciler.go
Comment (vaibhavjainwiz) [reply controllers/reconcilers/kserve_networkpolicy_reconciler.go:64]: We already have a ticket to make hard-coded configuration as configurable. I am adding this one to that ticket itself.
https://github.com/opendatahub-io/odh-model-controller/issues/85

**Comment by vaibhavjainwiz** [comment]
> There are some files missing the license headers while others have. Do we want to keep it aligned?

done

File: controllers/reconcilers/reconciler.go
Comment (vaibhavjainwiz) [reply controllers/reconcilers/reconciler.go:1]: done.. added it back

File: controllers/reconcilers/mm_route_reconciler.go
Comment (vaibhavjainwiz) [reply controllers/reconcilers/mm_route_reconciler.go]: done

**Review by davidesalerno — APPROVED** [review]

**Review by spolti — APPROVED** [review]

**Comment by spolti** [comment]
before merge we might need to fix the failing test

**Comment by Jooho** [comment]
@spolti @davidesalerno could you please share how you test this? I followed the description and it failed to deploy odh-model-controller image. Moreover, it used a stable image, not the image built for this PR.

**Comment by vaibhavjainwiz** [comment]
> @spolti @davidesalerno could you please share how you test this? I followed the description and it failed to deploy odh-model-controller image. Moreover, it used a stable image, not the image built for this PR.

@davidesalerno has faced the similiar issue. `DSCInitialization` settings are not picked by v2 operator automatically so we have to apply manifest manually and then it works.

**Comment by Jooho** [comment]
@vaibhavjainwiz @spolti i couldnt test this pr. @spolti please do not leave `lgtm` before testing. Openshif ci will merge if there is A `lgtm` on the PR. 
@vaibhavjainwiz please update description for test steps properly.

**Comment by vaibhavjainwiz** [comment]
> @vaibhavjainwiz @spolti i couldnt test this pr. @spolti please do not leave `lgtm` before testing. Openshif ci will merge if there is A `lgtm` on the PR. @vaibhavjainwiz please update description for test steps properly.

@Jooho I had updated the description for test. Please let me know if you are facing any issue to setup env.

**Pull request merged**
